protocol/protocol.go

```diff
@@ -0,0 +1,687 @@
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	signedResponseBytes, ok := reply[tagSREP]
+	if !ok {
+		return mjd.Mjd{}, 0, errors.New("protocol: response is missing signed portion")
+	}
+
+	if !ed25519.Verify(delegatedPublicKey, []byte(signedResponseContext+string(signedResponseBytes)), responseSigBytes) {
+		return mjd.Mjd{}, 0, errors.New("protocol: invalid response signature")
+	}
+
+	signedResponse, err := Decode(signedResponseBytes)
+	if err != nil {
+		return mjd.Mjd{}, 0, errors.New("protocol: failed to parse signed response: " + err.Error())
+	}
+
+	root, err := getFixedLength(signedResponse, tagROOT, "root", 32)
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	midpoint, err := getTimestamp(signedResponse, tagMIDP, "midpoint")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	radius, err = getUint32(signedResponse, tagRADI, "radius")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	// We now need to do some arithmetic.
+
+	if maxTime.Cmp(minTime) < 0 {
+		return mjd.Mjd{}, 0, errors.New("protocol: invalid delegation range")
+	}
+
+	if midpoint.Cmp(minTime) < 0 || maxTime.Cmp(midpoint) < 0 {
+		return mjd.Mjd{}, 0, errors.New("protocol: timestamp out of range for delegation")
+	}
+
+	index, err := getUint32(reply, tagINDX, "index")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+
+	path, err := getValue(reply, tagPATH, "path")
+	if err != nil {
+		return mjd.Mjd{}, 0, err
+	}
+	if len(path)%32 != 0 {
+		return mjd.Mjd{}, 0, errors.New("protocol: path is not a multiple of 32")
+	}
+
+	var hash [sha512.Size]byte
+	hashLeaf(&hash, nonce[:])
+
+	for len(path) > 0 {
+		pathElementIsRight := index&1 == 0
+		if pathElementIsRight {
+			hashNode(&hash, hash[:32], path[:32])
+		} else {
+			hashNode(&hash, path[:32], hash[:32])
+		}
+
+		index >>= 1
+		path = path[32:]
+	}
+
+	if !bytes.Equal(hash[:32], root) {
+		return mjd.Mjd{}, 0, errors.New("protocol: calculated tree root doesn't match signed root")
+	}
+
+	return midpoint, radius, nil
+}
+
+// EncapsulatePacket creates a UDP/TCP payload containing a roughtime message.
+func EncapsulatePacket(version uint32, message []byte) []byte {
+	length := len(message)
+	ret := make([]byte, length+12)
+	copy(ret, "ROUGHTIM")
+	binary.LittleEndian.PutUint32(ret[8:], uint32(length))
+	copy(ret[12:], message)
+	return ret
+}
+
+// DencapsulatePacket removes the encapsulation layer.
+func DencapsulatePacket(packet []byte) ([]byte, error) {
+	if bytes.Compare(packet[0:8], []byte("ROUGHTIM")) != 0 {
+		return nil, errors.New("Header invalid")
+	}
+	length := int(binary.LittleEndian.Uint32(packet[8:12]))
+	if length != len(packet)-12 {
+		return nil, errors.New("Mangled length!")
+	}
+	return packet[12:], nil
+
+}
+
+func isCompatibleVersion(list []byte, version uint32) bool {
+	if len(list)%4 != 0 {
+		return false
+	}
+
+	for ptr := 0; ptr < len(list); ptr += 4 {
+		if binary.LittleEndian.Uint32(list[ptr:ptr+4]) == version {
```

Comment:
just double-check, whether this function only checks that one in the list has the same version, or it must check that everything in the list matches the same version?

Reply:
The list needs to have our version in it. Other versions are irrelevant.